### PR TITLE
CTB: Dynamic zones and nested components support 

### LINKED ISSUE #2497: Split the frontend date picker into 3 new types; date, datetime, and time
- [x] **I have created my request on the Product Board before I submitted this issue**
- [x] **I have looked at all the other requests on the Product Board before I submitted this issue**

**Please describe your feature request:**

Using the following react frontend parts I think we can easily split the current date field option into these three:

- date (2018/12/30) => https://www.npmjs.com/package/react-date-picker
- datetime (2018/12/30 01:02:03) => https://www.npmjs.com/package/react-datetime-picker
- time (01:02:03) => https://www.npmjs.com/package/react-time-picker

This FR falls in line with the current focus on bookshelf date functions: https://github.com/strapi/strapi/issues/2495

### LINKED ISSUE #3716: CamelCase model names makes strapi unusable.
**Describe the bug**
I can not use camel case (or underscores or dashes) in model names to separate words in multi-word name.

I have 2 models - Question and QuestionType and I want to create a one-to-many relationship between them.

When I generate `strapi generate:api QuestionType` all seems working. Problems start when I add a relation field to QuestionType.questions (many : 1 ) Question.questionType.
It finishes without issues but then: 

1. it breaks admin UI. See screenshot (it renamed QuestionType to Questiontype and added questionType which is a field!, not a content type). Notice inconsistency that "Question Type" in Content type builder is still correct (but with space - which I think is nice formatting but adds more to confusion).

2. "QuestionType" becomes unusable in admin UI. When I try to add one record, screen loads indefinitely. (see screenshot 2)

(3. with graphql plugin, this will crash the server so I had to uninstall it)

(4. generate:api allows using CamelCase for model names, but Admin UI doesn't - it forces you to all lowercase first time, but you can rename it later - which causes same issues as when generating.

**Steps to reproduce the behavior**
1. Create Question type in UI  with string field "name"
2. generate QuestionType api `strapi generate:api QuestionType` or create it in UI with lowercase and then rename it to QuestionType (because of the bug 4.)
3. in UI create relation field QuestionType.questions (many : 1 ) Question.questionType. (see screenshot
4. in UI try to add new Quest

**Expected behavior**
I can name my models with camel case (or underscores or dashes to make multi-word models clearly understandable) and be able to use relations and clear UI.

**Screenshots**
scr1
scr2
scr3

**System**
- Node.js version: 10.15.1
- NPM version: 6.8.0
- Strapi version: Strapi v3.0.0-beta.14
- Database: PostgreSQL
- Operating system: MacOS 10.14.6

### LINKED ISSUE #3998: In Repeatable groups with relation field the description of the field is not aligned with non relation fields
**Describe the bug**
In Repeatable groups with relation field the description of the field is not aligned with non relation fields.

see below image

**Steps to reproduce the behavior**
1. Create Group with a number field and a relation field
2. Go to Content type manage and add a description to the number field and to the relation field
3. Add the group to a content type
4. Go to content type page
5. Click on create new content type
6. Go to the section of the repeatable group

**Expected behavior**
Description texts should be aligned either above the input field or below it

**Screenshots**
image

**System**
- Node.js version: v12
- Strapi version: beta.16.3

### LINKED ISSUE #4273: Add save button on top and bottom of create an entry
- [x] **I have created my request on the Product Board before I submitted this issue**
- [x] **I have looked at all the other requests on the Product Board before I submitted this issue**

**Please describe your feature request:**
On the create an entry page the save/reset button should be on the top and bottom for content types that have a lot of different fields so you don't have to scroll back up. A floating save button could also work.

### LINKED ISSUE #4673: Unreadable Field Descriptions (Div Element Overlap)
**Describe the bug**
Descriptions of the Fields in the Content Type detail are overlapping and text is unreadable.

**Steps to reproduce the behavior**
1. Go to Content Type Manager
2. Click on 'any content type
3. Paste the long text into description field
4. Repeat it for next field
5. Align these items into columns
6. Save
7. Go to Content Type
8. Select any entity
9. See error

**Expected behavior**
The description should be aligned into the contained element.

**Screenshots**
image

**System**
- Node.js version: 
- NPM version: 10
- Strapi version: v3.0.0-beta-17.8
- Database: postgress

### LINKED ISSUE #4707: Admin development server host config option
I think this feature request is too small to put it on ProductBoard, so here I go:

I'd like to have an option to define the listening `host` of the development server on `strapi develop --watch-admin`. This is useful when using a Docker-based setup. Right now, I can't access the development server as it's hardcoded to `localhost`.

### PR DESCRIPTION
**Pull request by soupette**
#### Description of what you did:
This PR makes the content types builder plugin support the dynamic zones and the nested components feature. The plugin has been recoded (a bit) with the simple idea the data management should be separated as much as possible from the UI one to allow easier customisation in the future.

**NOTE**: The translation files need to be cleaned but it will come with another PR.

- Fixes #2497
- Fixes #3716 
- Fixes #3998
- Fixes #4273
- Fixes #4673
- Fixes #4707 
#### My PR is a:

- [x] 💥 Breaking change
- [X] 🐛 Bug fix
- [X] 💅 Enhancement
- [x] 🚀 New feature

#### Main update on the:

- [ ] Admin
- [ ] Documentation
- [ ] Framework
- [x] Plugin

#### Manual testing done on the following databases:

- [ ] Not applicable
- [ ] MongoDB
- [ ] MySQL
- [ ] Postgres
- [ ] SQLite

### CODE REVIEW

File: packages/strapi-plugin-content-manager/admin/src/containers/ListView/index.js [hunk 1]
```diff
@@ -132,7 +131,8 @@ function ListView({
   const getListSchema = () => get(layouts, [...contentTypePath, 'schema'], {});
 
   const getName = () => {
-    return getFeatureLabel(models, slug);
+    // return getFeatureLabel(models, slug);
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/containers/ListView/index.js]: Remove the comment

File: packages/strapi-plugin-content-manager/admin/src/utils/getFeatureLabel.js [hunk 2]
```diff
@@ -1,5 +1,6 @@
 const getFeatureLabel = (data, slug) => {
   const { label } = data.find(item => item.uid === slug);
+
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/utils/getFeatureLabel.js]: This file is not needed anymore it needs to be removed

File: packages/strapi-plugin-content-type-builder/admin/src/containers/DataManagerProvider/reducer.js [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+  initialComponents: {},
+  intialContentTypes: {},
+  initialData: {},
+  modifiedData: {},
+  isLoading: true,
+  isLoadingForDataToBeSet: true,
+});
+
+const ONE_SIDE_RELATIONS = ['oneWay', 'manyWay'];
+const getOppositeNature = originalNature => {
```
Comment (soupette) [inline packages/strapi-plugin-content-type-builder/admin/src/containers/DataManagerProvider/reducer.js]: This should be a util

File: packages/strapi-plugin-content-type-builder/admin/src/containers/DataManagerProvider/reducer.js [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+            fromJS(newSchema)
+          );
+      }
+
+      return state.updateIn(['components', action.uid], () =>
+        fromJS(newSchema)
+      );
+    }
+    case 'DELETE_NOT_SAVED_TYPE': {
+      // Doing so will also reset the modified initial data
```
Comment (soupette) [inline packages/strapi-plugin-content-type-builder/admin/src/containers/DataManagerProvider/reducer.js]: Doing so will also reset the modified and the initial data

File: packages/strapi-plugin-content-type-builder/admin/src/containers/DataManagerProvider/reducer.js [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+      const pathToAttributeToRemove = [
+        ...pathToAttributes,
+        attributeToRemoveName,
+      ];
+
+      const attributeToRemoveData = state.getIn(pathToAttributeToRemove);
+
+      const isRemovingRelationAttribute =
+        attributeToRemoveData.get('nature') !== undefined;
+      // Only content types can have relations that with themselves since
```
Comment (soupette) [inline packages/strapi-plugin-content-type-builder/admin/src/containers/DataManagerProvider/reducer.js]: relations with

File: packages/strapi-plugin-content-type-builder/admin/src/containers/DataManagerProvider/utils/createModifiedDataSchema.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  const keyName = isInContentTypeView ? 'contentType' : 'component';
+  const schema = {
+    [keyName]: contentTypeSchema,
+    components: componentsAssociatedToContentType,
+  };
+
+  return schema;
+};
+
+const orderAllDataAttributesWithImmutable = (
```
Comment (soupette) [inline packages/strapi-plugin-content-type-builder/admin/src/containers/DataManagerProvider/utils/createModifiedDataSchema.js]: This needs to be improved in another PR

File: packages/strapi-plugin-content-type-builder/admin/src/containers/FormModal/index.js [hunk 7]
```diff
@@ -0,0 +1,1346 @@
...
+  } = useDataManager();
+  const {
+    componentToCreate,
+    formErrors,
+    initialData,
+    isCreatingComponentWhileAddingAField,
+    modifiedData,
+  } = reducerState.toJS();
+
+  // TODO close the modal when the user tries to create a field and reloaded the app
```
Comment (soupette) [inline packages/strapi-plugin-content-type-builder/admin/src/containers/FormModal/index.js]: Remove the comment is has been developed

File: packages/strapi-plugin-content-type-builder/admin/src/containers/FormModal/index.js [hunk 8]
```diff
@@ -0,0 +1,1346 @@
...
+      }
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [search]);
+
+  const form = get(forms, [state.modalType, 'form', state.settingType], () => ({
+    items: [],
+  }));
+
+  // TODO improve icon logic
```
Comment (soupette) [inline packages/strapi-plugin-content-type-builder/admin/src/containers/FormModal/index.js]: The icon type should be store in the state so it's easier to modify the stepper.
This needs to be added in another PR

File: packages/strapi-plugin-content-type-builder/admin/src/containers/FormModal/index.js [hunk 9]
```diff
@@ -0,0 +1,1346 @@
...
+  };
+  const handleSubmit = async (e, shouldContinue = isCreating) => {
+    e.preventDefault();
+
+    try {
+      await checkFormValidity();
+      const targetUid =
+        state.forTarget === 'components' ? state.targetUid : uid;
+
+      // TODO REMOVE and use makeSearch
```
Comment (soupette) [inline packages/strapi-plugin-content-type-builder/admin/src/containers/FormModal/index.js]: This needs be done

File: packages/strapi-plugin-content-type-builder/admin/src/containers/FormModal/index.js [hunk 10]
```diff
@@ -0,0 +1,1346 @@
...
+              // either 'EDIT_ATTRIBUTE' or 'ADD_ATTRIBUTE' in the DataManagerProvider
+              state.actionType === 'edit',
+              // This is for the edit part
+              initialData,
+              // Passing true will add the component to the components object
+              // This way we can add fields to the added component (if it wasn't there already)
+              true
+            );
+
+            // TODO change the search so the modal header is kept
```
Comment (soupette) [inline packages/strapi-plugin-content-type-builder/admin/src/containers/FormModal/index.js]: This needs be done

File: packages/strapi-plugin-content-type-builder/admin/src/containers/FormModal/utils/componentForm.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    const items = [
+      [
+        {
+          autoFocus: true,
+          name: `${prefix}name`,
+          type: 'text',
+          label: {
+            id: getTrad('modalForm.attribute.form.base.name'),
+          },
+          // description: {
```
Comment (soupette) [inline packages/strapi-plugin-content-type-builder/admin/src/containers/FormModal/utils/componentForm.js]: This needs to be removed

File: packages/strapi-plugin-content-type-builder/admin/src/containers/ListView/index.js [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+    className: 'button-secondary',
+  };
+
+  const listActions = isInDevelopmentMode
+    ? [{ ...configureButtonProps }, { ...addButtonProps }]
+    : [configureButtonProps];
+
+  const handleClickOnTrashIcon = () => {};
+
+  const CustomRow = props => {
```
Comment (soupette) [inline packages/strapi-plugin-content-type-builder/admin/src/containers/ListView/index.js]: This should be a component in a separated file

File: packages/strapi-plugin-content-type-builder/admin/src/components/ComponentCard/index.js [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    </Wrapper>
+  );
+}
+
+ComponentCard.defaultProps = {
+  component: null,
+  isActive: false,
+  isInDevelopmentMode: false,
+  onClick: () => {},
+  onRemoveClick: () => {},
```
Comment (virginieky) [inline packages/strapi-plugin-content-type-builder/admin/src/components/ComponentCard/index.js]: `onRemoveClick` can be deleted, never used

File: packages/strapi-plugin-content-type-builder/admin/src/components/ComponentIconPicker/Wrapper.js [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+      if (error) {
+        return `
+          border: 1px solid ${colors.darkOrange};
+          border-radius: 2px;
+        `;
+      }
+    }}
+
+    &:focus,
+    :active {
```
Comment (virginieky) [inline packages/strapi-plugin-content-type-builder/admin/src/components/ComponentIconPicker/Wrapper.js]: `&:focus,
&:active`

File: packages/strapi-plugin-content-type-builder/admin/src/components/CreatableSelect/index.js [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+      // menuIsOpen
+      onChange={handleChange}
+      styles={styles}
+      options={formatOptions()}
+    />
+  );
+};
+
+CreatableSelect.defaultProps = {
+  error: null,
```
Comment (virginieky) [inline packages/strapi-plugin-content-type-builder/admin/src/components/CreatableSelect/index.js]: `error` is never used

File: packages/strapi-plugin-content-type-builder/admin/src/components/CreatableSelect/index.js [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+  const getValue = () => {
+    return formatOptions().find(el => el.value === value);
+  };
+
+  return (
+    <Creatable
+      defaultValue={getValue()}
+      isClearable
+      // menuIsOpen
```
Comment (virginieky) [inline packages/strapi-plugin-content-type-builder/admin/src/components/CreatableSelect/index.js]: should `menuIsOpen` prop be removed?

File: packages/strapi-plugin-content-type-builder/admin/src/components/DynamicZoneList/ComponentButton.js [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+      width: 11px;
+      height: 11px;
+    }
+  }
+  p {
+    font-size: 13px;
+    font-weight: bold;
+    color: #2c3138;
+    line-height: normal;
+    // margin-top: 7px;
```
Comment (virginieky) [inline packages/strapi-plugin-content-type-builder/admin/src/components/DynamicZoneList/ComponentButton.js]: This can be removed too

File: packages/strapi-helper-plugin/lib/src/components/Modal/index.js [hunk 18]
```diff
@@ -16,8 +16,8 @@ function WrapperModal({ children, isOpen, onToggle, ...rest }) {
   return (
     <Wrapper>
       <StyledModal isOpen={isOpen} toggle={onToggle} {...rest}>
-        <HeaderModal toggle={onToggle} />
-        <Close />
+        <HeaderModal toggle={onToggle} style={{ display: 'none' }} />
```
[the remark this answers is not in the document]
Comment (soupette) [reply packages/strapi-helper-plugin/lib/src/components/Modal/index.js]: We need that one it's hidden because we use close component
